File: ios/AppSettings.swift
import Foundation
import Combine
import SwiftUI

/// Up to 30 macro buttons (btn1…btn30). Receivers map these to gamepad buttons.
let kMaxButtons = 30

enum ButtonMode: Int, CaseIterable, Identifiable {
    case hold = 0, toggle = 1, tap = 2
    var id: Int { rawValue }
    var label: String {
        switch self {
        case .hold:   return "Hold"
        case .toggle: return "Toggle"
        case .tap:    return "Tap"
        }
    }
}

// MARK: - Import/Export Profile
struct LayoutProfile: Codable {
    var buttonCount: Int
    var buttonLabels: [String]
    var buttonColors: [Int]
    var buttonModes: [Int]
    var buttonPosX: [Double]
    var buttonPosY: [Double]
    var buttonSize: [Double]
    var showWheel: Bool
    var wheelPosX: Double
    var wheelPosY: Double
    var wheelSize: Double
    var sliderPosX: Double
    var sliderPosY: Double
    var sliderWidth: Double
    var sliderHeight: Double
}

/// User-tunable configuration + freeform HUD layout, persisted to UserDefaults.
/// Positions are normalized (0…1) to the play area so they adapt to any device.
final class AppSettings: ObservableObject {
    private let store = UserDefaults.standard

    // MARK: Connection
    @Published var host: String        { didSet { store.set(host, forKey: "host") } }
    @Published var port: Int           { didSet { store.set(port, forKey: "port") } }
    @Published var autoConnect: Bool   { didSet { store.set(autoConnect, forKey: "autoConnect") } }
    /// Outgoing packet rate. 60 / 90 / 120.
    @Published var sendRateHz: Int     { didSet { store.set(sendRateHz, forKey: "sendRateHz") } }

    // MARK: Steering
    @Published var sensitivity: Double     { didSet { store.set(sensitivity, forKey: "sensitivity") } }
    @Published var deadzoneDegrees: Double { didSet { store.set(deadzoneDegrees, forKey: "deadzoneDegrees") } }
    @Published var maxLockDegrees: Double  { didSet { store.set(maxLockDegrees, forKey: "maxLockDegrees") } }
    @Published var smoothing: Double       { didSet { store.set(smoothing, forKey: "smoothing") } }
    @Published var steerCurve: Double      { didSet { store.set(steerCurve, forKey: "steerCurve") } }
    @Published var invertSteering: Bool    { didSet { store.set(invertSteering, forKey: "invertSteering") } }
    @Published var autoInvert: Bool        { didSet { store.set(autoInvert, forKey: "autoInvert") } }
    @Published var calibrateOnLaunch: Bool { didSet { store.set(calibrateOnLaunch, forKey: "calibrateOnLaunch") } }

    // MARK: Assists
    /// 0 = Off, 1 = Light, 2 = Strong. Setting it applies a preset.
    @Published var assistLevel: Int        { didSet { store.set(assistLevel, forKey: "assistLevel"); applyAssistPreset() } }
    /// Auto-trim: while held near center, slowly re-zeros so hand-shake on a
    /// straight doesn't slowly pull the car off line.
    @Published var recenterAssist: Bool    { didSet { store.set(recenterAssist, forKey: "recenterAssist") } }
    /// Max steering change per second (0 = off). Filters jitter spikes.
    @Published var steerRateLimit: Double  { didSet { store.set(steerRateLimit, forKey: "steerRateLimit") } }
    /// Throttle lift-off ease, in seconds (0 = instant). Throttle APPLIES
    /// instantly but RELEASES gradually when you lift your finger — like easing
    /// off the gas. Braking is never eased (instant release).
    @Published var throttleEase: Double  { didSet { store.set(throttleEase, forKey: "throttleEase") } }
    /// Pedal response curves (gamma). >1 = gentle at the start, steep at the end
    /// (trail-braking: ~1.9 maps 70% travel to ~50% force for precision near lockup).
    @Published var brakeGamma: Double      { didSet { store.set(brakeGamma, forKey: "brakeGamma") } }
    @Published var throttleGamma: Double   { didSet { store.set(throttleGamma, forKey: "throttleGamma") } }
    /// Vibrate on telemetry-detected lockup / wheelspin (cues sent by the Mac).
    @Published var telemetryHaptics: Bool  { didSet { store.set(telemetryHaptics, forKey: "telemetryHaptics") } }

    // MARK: Pedals
    @Published var invertPedals: Bool  { didSet { store.set(invertPedals, forKey: "invertPedals") } }

    // MARK: Feedback
    @Published var hapticsEnabled: Bool { didSet { store.set(hapticsEnabled, forKey: "hapticsEnabled") } }
    @Published var hapticOnConnect: Bool { didSet { store.set(hapticOnConnect, forKey: "hapticOnConnect") } }

    // MARK: Appearance
    @Published var accentColorIndex: Int   { didSet { store.set(accentColorIndex, forKey: "accentColorIndex") } }
    @Published var backgroundStyle: Int    { didSet { store.set(backgroundStyle, forKey: "backgroundStyle") } }
    @Published var controlOpacity: Double  { didSet { store.set(controlOpacity, forKey: "controlOpacity") } }
    @Published var buttonShape: Int        { didSet { store.set(buttonShape, forKey: "buttonShape") } }
    @Published var showAngleReadout: Bool  { didSet { store.set(showAngleReadout, forKey: "showAngleReadout") } }
    @Published var throttleColorIndex: Int { didSet { store.set(throttleColorIndex, forKey: "throttleColorIndex") } }
    @Published var brakeColorIndex: Int    { didSet { store.set(brakeColorIndex, forKey: "brakeColorIndex") } }

    // MARK: Onboarding
    @Published var hasOnboarded: Bool  { didSet { store.set(hasOnboarded, forKey: "hasOnboarded") } }

    // MARK: Macro buttons
    @Published var buttonCount: Int       { didSet { store.set(buttonCount, forKey: "buttonCount") } }
    @Published var buttonLabels: [String] { didSet { store.set(buttonLabels, forKey: "buttonLabels") } }
    @Published var buttonColors: [Int]    { didSet { store.set(buttonColors, forKey: "buttonColors") } }
    @Published var buttonModes: [Int]     { didSet { store.set(buttonModes, forKey: "buttonModes") } }
    @Published var buttonPosX: [Double]   { didSet { store.set(buttonPosX, forKey: "buttonPosX") } }
    @Published var buttonPosY: [Double]   { didSet { store.set(buttonPosY, forKey: "buttonPosY") } }
    @Published var buttonSize: [Double]   { didSet { store.set(buttonSize, forKey: "buttonSize") } }

    // MARK: HUD elements (normalized center + size)
    @Published var showWheel: Bool       { didSet { store.set(showWheel, forKey: "showWheel") } }
    @Published var wheelPosX: Double     { didSet { store.set(wheelPosX, forKey: "wheelPosX") } }
    @Published var wheelPosY: Double     { didSet { store.set(wheelPosY, forKey: "wheelPosY") } }
    @Published var wheelSize: Double     { didSet { store.set(wheelSize, forKey: "wheelSize") } }
    @Published var sliderPosX: Double    { didSet { store.set(sliderPosX, forKey: "sliderPosX") } }
    @Published var sliderPosY: Double    { didSet { store.set(sliderPosY, forKey: "sliderPosY") } }
    @Published var sliderWidth: Double   { didSet { store.set(sliderWidth, forKey: "sliderWidth") } }
    @Published var sliderHeight: Double  { didSet { store.set(sliderHeight, forKey: "sliderHeight") } }

    // MARK: Defaults
    static func defaultLabels() -> [String] {
        let named = ["A", "B", "X", "Y", "LB", "RB", "LT", "RT", "L3", "R3"]
        return (0..<kMaxButtons).map { $0 < named.count ? named[$0] : "\($0 + 1)" }
    }
    static func defaultColors() -> [Int] { (0..<kMaxButtons).map { $0 % 8 } }
    static func defaultModes() -> [Int]  { Array(repeating: 0, count: kMaxButtons) }
    // 5-column grid down the left side.
    static func defaultPosX() -> [Double] { (0..<kMaxButtons).map { 0.07 + Double($0 % 5) * 0.09 } }
    static func defaultPosY() -> [Double] { (0..<kMaxButtons).map { 0.28 + Double($0 / 5) * 0.12 } }
    static func defaultSizes() -> [Double] { Array(repeating: 72, count: kMaxButtons) }

    init() {
        host             = store.string(forKey: "host") ?? "192.168.1.50"
        port             = store.object(forKey: "port") as? Int ?? 5005
        autoConnect      = store.object(forKey: "autoConnect") as? Bool ?? false
        sendRateHz       = store.object(forKey: "sendRateHz") as? Int ?? 120
        sensitivity      = store.object(forKey: "sensitivity") as? Double ?? 1.0
        deadzoneDegrees  = store.object(forKey: "deadzoneDegrees") as? Double ?? 2.5
        maxLockDegrees   = store.object(forKey: "maxLockDegrees") as? Double ?? 90.0
        smoothing        = store.object(forKey: "smoothing") as? Double ?? 0.25
        steerCurve       = store.object(forKey: "steerCurve") as? Double ?? 1.0
        invertSteering   = store.object(forKey: "invertSteering") as? Bool ?? false
        autoInvert       = store.object(forKey: "autoInvert") as? Bool ?? true
        calibrateOnLaunch = store.object(forKey: "calibrateOnLaunch") as? Bool ?? true
        assistLevel      = store.object(forKey: "assistLevel") as? Int ?? 1
        recenterAssist   = store.object(forKey: "recenterAssist") as? Bool ?? true
        steerRateLimit   = store.object(forKey: "steerRateLimit") as? Double ?? 6.0
        throttleEase     = store.object(forKey: "throttleEase") as? Double ?? 0.25
        brakeGamma       = store.object(forKey: "brakeGamma") as? Double ?? 1.9
        throttleGamma    = store.object(forKey: "throttleGamma") as? Double ?? 1.0
        telemetryHaptics = store.object(forKey: "telemetryHaptics") as? Bool ?? true
        invertPedals     = store.object(forKey: "invertPedals") as? Bool ?? false
        hapticsEnabled   = store.object(forKey: "hapticsEnabled") as? Bool ?? true
        hapticOnConnect  = store.object(forKey: "hapticOnConnect") as? Bool ?? true
        accentColorIndex = store.object(forKey: "accentColorIndex") as? Int ?? 0
        backgroundStyle  = store.object(forKey: "backgroundStyle") as? Int ?? 1
        controlOpacity   = store.object(forKey: "controlOpacity") as? Double ?? 1.0
        buttonShape      = store.object(forKey: "buttonShape") as? Int ?? 0
        showAngleReadout = store.object(forKey: "showAngleReadout") as? Bool ?? true
        throttleColorIndex = store.object(forKey: "throttleColorIndex") as? Int ?? 0
        brakeColorIndex  = store.object(forKey: "brakeColorIndex") as? Int ?? 1
        hasOnboarded     = store.object(forKey: "hasOnboarded") as? Bool ?? false
        buttonCount      = store.object(forKey: "buttonCount") as? Int ?? 4
        buttonLabels     = AppSettings.pad(store.array(forKey: "buttonLabels") as? [String], AppSettings.defaultLabels())
        buttonColors     = AppSettings.pad(store.array(forKey: "buttonColors") as? [Int], AppSettings.defaultColors())
        buttonModes      = AppSettings.pad(store.array(forKey: "buttonModes") as? [Int], AppSettings.defaultModes())
        buttonPosX       = AppSettings.pad(store.array(forKey: "buttonPosX") as? [Double], AppSettings.defaultPosX())
        buttonPosY       = AppSettings.pad(store.array(forKey: "buttonPosY") as? [Double], AppSettings.defaultPosY())
        buttonSize       = AppSettings.pad(store.array(forKey: "buttonSize") as? [Double], AppSettings.defaultSizes())
        showWheel        = store.object(forKey: "showWheel") as? Bool ?? true
        wheelPosX        = store.object(forKey: "wheelPosX") as? Double ?? 0.5
        wheelPosY        = store.object(forKey: "wheelPosY") as? Double ?? 0.56
        wheelSize        = store.object(forKey: "wheelSize") as? Double ?? 180
        sliderPosX       = store.object(forKey: "sliderPosX") as? Double ?? 0.92
        sliderPosY       = store.object(forKey: "sliderPosY") as? Double ?? 0.5
        sliderWidth      = store.object(forKey: "sliderWidth") as? Double ?? 96
        sliderHeight     = store.object(forKey: "sliderHeight") as? Double ?? 320
    }

    /// Bundle steering assists into one easy knob.
    func applyAssistPreset() {
        switch assistLevel {
        case 0: // Off — direct
            smoothing = 0.10; deadzoneDegrees = 1.5; recenterAssist = false
            steerRateLimit = 0; throttleEase = 0
        case 2: // Strong — very forgiving
            smoothing = 0.45; deadzoneDegrees = 4.0; recenterAssist = true
            steerRateLimit = 3.5; throttleEase = 0.40
        default: // Light — recommended
            smoothing = 0.25; deadzoneDegrees = 2.5; recenterAssist = true
            steerRateLimit = 6.0; throttleEase = 0.25
        }
    }

    func resetLayout() {
        buttonPosX = AppSettings.defaultPosX()
        buttonPosY = AppSettings.defaultPosY()
        buttonSize = AppSettings.defaultSizes()
        showWheel = true
        wheelPosX = 0.5; wheelPosY = 0.56; wheelSize = 180
        sliderPosX = 0.92; sliderPosY = 0.5; sliderWidth = 96; sliderHeight = 320
    }

    func resetAll() {
        autoConnect = false; sendRateHz = 120
        sensitivity = 1.0; maxLockDegrees = 90.0; steerCurve = 1.0; invertSteering = false
        autoInvert = true; calibrateOnLaunch = true
        assistLevel = 1; applyAssistPreset()
        brakeGamma = 1.9; throttleGamma = 1.0; telemetryHaptics = true
        invertPedals = false; hapticsEnabled = true; hapticOnConnect = true
        accentColorIndex = 0; backgroundStyle = 1; controlOpacity = 1.0
        buttonShape = 0; showAngleReadout = true; throttleColorIndex = 0; brakeColorIndex = 1
        buttonCount = 4
        buttonLabels = AppSettings.defaultLabels()
        buttonColors = AppSettings.defaultColors()
        buttonModes  = AppSettings.defaultModes()
        resetLayout()
    }

    // MARK: - Import/Export Methods
    func exportLayout() -> String {
        let profile = LayoutProfile(
            buttonCount: buttonCount, buttonLabels: buttonLabels, buttonColors: buttonColors, buttonModes: buttonModes,
            buttonPosX: buttonPosX, buttonPosY: buttonPosY, buttonSize: buttonSize,
            showWheel: showWheel, wheelPosX: wheelPosX, wheelPosY: wheelPosY, wheelSize: wheelSize,
            sliderPosX: sliderPosX, sliderPosY: sliderPosY, sliderWidth: sliderWidth, sliderHeight: sliderHeight
        )
        if let data = try? JSONEncoder().encode(profile), let str = String(data: data, encoding: .utf8) {
            return str
        }
        return ""
    }

    func importLayout(json: String) -> Bool {
        guard let data = json.data(using: .utf8),
              let profile = try? JSONDecoder().decode(LayoutProfile.self, from: data) else { return false }
        
        self.buttonCount = profile.buttonCount
        self.buttonLabels = AppSettings.pad(profile.buttonLabels, AppSettings.defaultLabels())
        self.buttonColors = AppSettings.pad(profile.buttonColors, AppSettings.defaultColors())
        self.buttonModes = AppSettings.pad(profile.buttonModes, AppSettings.defaultModes())
        self.buttonPosX = AppSettings.pad(profile.buttonPosX, AppSettings.defaultPosX())
        self.buttonPosY = AppSettings.pad(profile.buttonPosY, AppSettings.defaultPosY())
        self.buttonSize = AppSettings.pad(profile.buttonSize, AppSettings.defaultSizes())
        
        self.showWheel = profile.showWheel
        self.wheelPosX = profile.wheelPosX
        self.wheelPosY = profile.wheelPosY
        self.wheelSize = profile.wheelSize
        
        self.sliderPosX = profile.sliderPosX
        self.sliderPosY = profile.sliderPosY
        self.sliderWidth = profile.sliderWidth
        self.sliderHeight = profile.sliderHeight
        return true
    }

    // Bindings into array elements (UIs use these so element edits persist).
    func labelBinding(_ i: Int) -> Binding<String> { Binding(get: { self.buttonLabels[i] }, set: { self.buttonLabels[i] = $0 }) }
    func colorBinding(_ i: Int) -> Binding<Int>    { Binding(get: { self.buttonColors[i] }, set: { self.buttonColors[i] = $0 }) }
    func modeBinding(_ i: Int) -> Binding<Int>     { Binding(get: { self.buttonModes[i] }, set: { self.buttonModes[i] = $0 }) }
    func sizeBinding(_ i: Int) -> Binding<Double>  { Binding(get: { self.buttonSize[i] }, set: { self.buttonSize[i] = $0 }) }
    func posXBinding(_ i: Int) -> Binding<Double>  { Binding(get: { self.buttonPosX[i] }, set: { self.buttonPosX[i] = $0 }) }
    func posYBinding(_ i: Int) -> Binding<Double>  { Binding(get: { self.buttonPosY[i] }, set: { self.buttonPosY[i] = $0 }) }
    var wheelXBinding: Binding<Double> { Binding(get: { self.wheelPosX }, set: { self.wheelPosX = $0 }) }
    var wheelYBinding: Binding<Double> { Binding(get: { self.wheelPosY }, set: { self.wheelPosY = $0 }) }
    var sliderXBinding: Binding<Double> { Binding(get: { self.sliderPosX }, set: { self.sliderPosX = $0 }) }
    var sliderYBinding: Binding<Double> { Binding(get: { self.sliderPosY }, set: { self.sliderPosY = $0 }) }

    private static func pad<T>(_ arr: [T]?, _ def: [T]) -> [T] {
        var a = arr ?? def
        if a.count < kMaxButtons { a += Array(def[a.count..<kMaxButtons]) }
        return Array(a.prefix(kMaxButtons))
    }
}

File: android/app/src/main/java/com/gyrowheel/AppSettings.kt
package com.gyrowheel

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.mutableStateOf
import org.json.JSONArray
import org.json.JSONObject
import kotlin.reflect.KProperty

/**
 * User-tunable configuration + freeform HUD layout, persisted to SharedPreferences.
 * Positions are normalized (0…1) to the play area so they adapt to any device.
 * Each property is backed by Compose state so the UI recomposes on change.
 */
class AppSettings(context: Context) {
    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences("gyrowheel.settings", Context.MODE_PRIVATE)

    // MARK: - Compose-observable, prefs-backed delegate
    private inner class Pref<T>(
        private val key: String,
        default: T,
        private val load: (String, T) -> T,
        private val save: (String, T) -> Unit
    ) {
        private val state = mutableStateOf(load(key, default))
        operator fun getValue(thisRef: Any?, p: KProperty<*>): T = state.value
        operator fun setValue(thisRef: Any?, p: KProperty<*>, v: T) {
            state.value = v
            save(key, v)
        }
    }

    private fun boolPref(key: String, def: Boolean) = Pref(key, def,
        { k, d -> prefs.getBoolean(k, d) }, { k, v -> prefs.edit().putBoolean(k, v).apply() })

    private fun intPref(key: String, def: Int) = Pref(key, def,
        { k, d -> prefs.getInt(k, d) }, { k, v -> prefs.edit().putInt(k, v).apply() })

    private fun dblPref(key: String, def: Double) = Pref(key, def,
        { k, d -> if (prefs.contains(k)) prefs.getFloat(k, d.toFloat()).toDouble() else d },
        { k, v -> prefs.edit().putFloat(k, v.toFloat()).apply() })

    private fun strPref(key: String, def: String) = Pref(key, def,
        { k, d -> prefs.getString(k, d) ?: d }, { k, v -> prefs.edit().putString(k, v).apply() })

    private fun intListPref(key: String, def: List<Int>) = Pref(key, def,
        { k, d -> prefs.getString(k, null)?.let { jsonToIntList(it) } ?: d },
        { k, v -> prefs.edit().putString(k, intListToJson(v)).apply() })

    private fun dblListPref(key: String, def: List<Double>) = Pref(key, def,
        { k, d -> prefs.getString(k, null)?.let { jsonToDblList(it) } ?: d },
        { k, v -> prefs.edit().putString(k, dblListToJson(v)).apply() })

    private fun strListPref(key: String, def: List<String>) = Pref(key, def,
        { k, d -> prefs.getString(k, null)?.let { jsonToStrList(it) } ?: d },
        { k, v -> prefs.edit().putString(k, strListToJson(v)).apply() })

    // MARK: - Connection
    var host by strPref("host", "192.168.1.50")
    var port by intPref("port", 5005)
    var autoConnect by boolPref("autoConnect", false)
    var sendRateHz by intPref("sendRateHz", 120)

    // MARK: - Steering
    var sensitivity by dblPref("sensitivity", 1.0)
    var deadzoneDegrees by dblPref("deadzoneDegrees", 2.5)
    var maxLockDegrees by dblPref("maxLockDegrees", 90.0)
    var smoothing by dblPref("smoothing", 0.25)
    var steerCurve by dblPref("steerCurve", 1.0)
    var invertSteering by boolPref("invertSteering", false)
    var autoInvert by boolPref("autoInvert", true)
    var calibrateOnLaunch by boolPref("calibrateOnLaunch", true)

    // MARK: - Assists
    var assistLevel by intPref("assistLevel", 1)
    var recenterAssist by boolPref("recenterAssist", true)
    var steerRateLimit by dblPref("steerRateLimit", 6.0)
    var throttleEase by dblPref("throttleEase", 0.25)
    var brakeGamma by dblPref("brakeGamma", 1.9)
    var throttleGamma by dblPref("throttleGamma", 1.0)
    var telemetryHaptics by boolPref("telemetryHaptics", true)

    // MARK: - Paddle Shifter (back-tap)
    var backTapShiftEnabled by boolPref("backTapShiftEnabled", false)
    var backTapSensitivity by dblPref("backTapSensitivity", 2.5) // m/s² per sample (Android units)
    var backTapSharpness by dblPref("backTapSharpness", 3.5)
    var upshiftButton by intPref("upshiftButton", 0)
    var downshiftButton by intPref("downshiftButton", 1)
    var shiftHaptics by boolPref("shiftHaptics", true)

    // MARK: - Pedals
    var invertPedals by boolPref("invertPedals", false)

    // MARK: - Feedback
    var hapticsEnabled by boolPref("hapticsEnabled", true)
    var hapticOnConnect by boolPref("hapticOnConnect", true)

    // MARK: - Appearance
    var accentColorIndex by intPref("accentColorIndex", 0)
    var backgroundStyle by intPref("backgroundStyle", 1)
    var controlOpacity by dblPref("controlOpacity", 1.0)
    var buttonShape by intPref("buttonShape", 0)
    var showAngleReadout by boolPref("showAngleReadout", true)
    var throttleColorIndex by intPref("throttleColorIndex", 0)
    var brakeColorIndex by intPref("brakeColorIndex", 1)

    // MARK: - Onboarding
    var hasOnboarded by boolPref("hasOnboarded", false)

    // MARK: - Macro buttons
    var buttonCount by intPref("buttonCount", 4)
    var buttonLabels by strListPref("buttonLabels", defaultLabels())
    var buttonColors by intListPref("buttonColors", defaultColors())
    var buttonModes by intListPref("buttonModes", defaultModes())
    var buttonPosX by dblListPref("buttonPosX", defaultPosX())
    var buttonPosY by dblListPref("buttonPosY", defaultPosY())
    var buttonSize by dblListPref("buttonSize", defaultSizes())

    // MARK: - HUD elements (normalized center + size)
    var showWheel by boolPref("showWheel", true)
    var wheelPosX by dblPref("wheelPosX", 0.5)
    var wheelPosY by dblPref("wheelPosY", 0.56)
    var wheelSize by dblPref("wheelSize", 180.0)
    var sliderPosX by dblPref("sliderPosX", 0.92)
    var sliderPosY by dblPref("sliderPosY", 0.5)
    var sliderWidth by dblPref("sliderWidth", 96.0)
    var sliderHeight by dblPref("sliderHeight", 320.0)

    // MARK: - Element mutators (reassign whole list so Compose + prefs update)
    fun setLabel(i: Int, v: String) { buttonLabels = buttonLabels.toMutableList().also { it[i] = v } }
    fun setColor(i: Int, v: Int) { buttonColors = buttonColors.toMutableList().also { it[i] = v } }
    fun setMode(i: Int, v: Int) { buttonModes = buttonModes.toMutableList().also { it[i] = v } }
    fun setSize(i: Int, v: Double) { buttonSize = buttonSize.toMutableList().also { it[i] = v } }
    fun setPosX(i: Int, v: Double) { buttonPosX = buttonPosX.toMutableList().also { it[i] = v } }
    fun setPosY(i: Int, v: Double) { buttonPosY = buttonPosY.toMutableList().also { it[i] = v } }

    /** Bundle steering assists into one easy knob. */
    fun applyAssistPreset() {
        when (assistLevel) {
            0 -> { smoothing = 0.10; deadzoneDegrees = 1.5; recenterAssist = false; steerRateLimit = 0.0; throttleEase = 0.0 }
            2 -> { smoothing = 0.45; deadzoneDegrees = 4.0; recenterAssist = true; steerRateLimit = 3.5; throttleEase = 0.40 }
            else -> { smoothing = 0.25; deadzoneDegrees = 2.5; recenterAssist = true; steerRateLimit = 6.0; throttleEase = 0.25 }
        }
    }

    fun resetLayout() {
        buttonPosX = defaultPosX(); buttonPosY = defaultPosY(); buttonSize = defaultSizes()
        showWheel = true
        wheelPosX = 0.5; wheelPosY = 0.56; wheelSize = 180.0
        sliderPosX = 0.92; sliderPosY = 0.5; sliderWidth = 96.0; sliderHeight = 320.0
    }

    fun resetAll() {
        autoConnect = false; sendRateHz = 120
        sensitivity = 1.0; maxLockDegrees = 90.0; steerCurve = 1.0; invertSteering = false
        autoInvert = true; calibrateOnLaunch = true
        assistLevel = 1; applyAssistPreset()
        brakeGamma = 1.9; throttleGamma = 1.0; telemetryHaptics = true
        backTapShiftEnabled = false; backTapSensitivity = 2.5; backTapSharpness = 3.5
        upshiftButton = 0; downshiftButton = 1; shiftHaptics = true
        invertPedals = false; hapticsEnabled = true; hapticOnConnect = true
        accentColorIndex = 0; backgroundStyle = 1; controlOpacity = 1.0
        buttonShape = 0; showAngleReadout = true; throttleColorIndex = 0; brakeColorIndex = 1
        buttonCount = 4
        buttonLabels = defaultLabels(); buttonColors = defaultColors(); buttonModes = defaultModes()
        resetLayout()
    }

    // MARK: - Import / Export (JSON keys match the iOS LayoutProfile for cross-device profiles)
    fun exportLayout(): String {
        val o = JSONObject()
        o.put("buttonCount", buttonCount)
        o.put("buttonLabels", JSONArray(buttonLabels))
        o.put("buttonColors", JSONArray(buttonColors))
        o.put("buttonModes", JSONArray(buttonModes))
        o.put("buttonPosX", JSONArray(buttonPosX))
        o.put("buttonPosY", JSONArray(buttonPosY))
        o.put("buttonSize", JSONArray(buttonSize))
        o.put("showWheel", showWheel)
        o.put("wheelPosX", wheelPosX); o.put("wheelPosY", wheelPosY); o.put("wheelSize", wheelSize)
        o.put("sliderPosX", sliderPosX); o.put("sliderPosY", sliderPosY)
        o.put("sliderWidth", sliderWidth); o.put("sliderHeight", sliderHeight)
        return o.toString()
    }

    fun importLayout(json: String): Boolean {
        return try {
            val o = JSONObject(json)
            buttonCount = o.optInt("buttonCount", buttonCount).coerceIn(2, kMaxButtons)
            buttonLabels = pad(jsonArrToStr(o.optJSONArray("buttonLabels")), defaultLabels())
            buttonColors = pad(jsonArrToInt(o.optJSONArray("buttonColors")), defaultColors())
            buttonModes = pad(jsonArrToInt(o.optJSONArray("buttonModes")), defaultModes())
            buttonPosX = pad(jsonArrToDbl(o.optJSONArray("buttonPosX")), defaultPosX())
            buttonPosY = pad(jsonArrToDbl(o.optJSONArray("buttonPosY")), defaultPosY())
            buttonSize = pad(jsonArrToDbl(o.optJSONArray("buttonSize")), defaultSizes())
            showWheel = o.optBoolean("showWheel", showWheel)
            wheelPosX = o.optDouble("wheelPosX", wheelPosX)
            wheelPosY = o.optDouble("wheelPosY", wheelPosY)
            wheelSize = o.optDouble("wheelSize", wheelSize)
            sliderPosX = o.optDouble("sliderPosX", sliderPosX)
            sliderPosY = o.optDouble("sliderPosY", sliderPosY)
            sliderWidth = o.optDouble("sliderWidth", sliderWidth)
            sliderHeight = o.optDouble("sliderHeight", sliderHeight)
            true
        } catch (e: Exception) {
            false
        }
    }

    companion object {
        fun defaultLabels(): List<String> {
            val named = listOf("A", "B", "X", "Y", "LB", "RB", "LT", "RT", "L3", "R3")
            return (0 until kMaxButtons).map { if (it < named.size) named[it] else "${it + 1}" }
        }
        fun defaultColors(): List<Int> = (0 until kMaxButtons).map { it % 8 }
        fun defaultModes(): List<Int> = List(kMaxButtons) { 0 }
        fun defaultPosX(): List<Double> = (0 until kMaxButtons).map { 0.07 + (it % 5) * 0.09 }
        fun defaultPosY(): List<Double> = (0 until kMaxButtons).map { 0.28 + (it / 5) * 0.12 }
        fun defaultSizes(): List<Double> = List(kMaxButtons) { 72.0 }

        fun <T> pad(arr: List<T>?, def: List<T>): List<T> {
            val a = (arr ?: def).toMutableList()
            if (a.size < kMaxButtons) a.addAll(def.subList(a.size, kMaxButtons))
            return a.take(kMaxButtons)
        }

        private fun jsonToIntList(s: String) = jsonArrToInt(JSONArray(s))
        private fun jsonToDblList(s: String) = jsonArrToDbl(JSONArray(s))
        private fun jsonToStrList(s: String) = jsonArrToStr(JSONArray(s))
        private fun intListToJson(v: List<Int>) = JSONArray(v).toString()
        private fun dblListToJson(v: List<Double>) = JSONArray(v).toString()
        private fun strListToJson(v: List<String>) = JSONArray(v).toString()

        private fun jsonArrToInt(a: JSONArray?): List<Int> =
            if (a == null) emptyList() else (0 until a.length()).map { a.getInt(it) }
        private fun jsonArrToDbl(a: JSONArray?): List<Double> =
            if (a == null) emptyList() else (0 until a.length()).map { a.getDouble(it) }
        private fun jsonArrToStr(a: JSONArray?): List<String> =
            if (a == null) emptyList() else (0 until a.length()).map { a.getString(it) }
    }
}
